ge/org.osate.ge/src/org/osate/ge/internal/ui/editor/actions/PasteAction.java

```diff
@@ -0,0 +1,471 @@
+				getWorkbenchPart().getAdapter(ProjectProvider.class), "Unable to retrieve project provider");
+		final ExtensionService extService = Objects.requireNonNull(
+				getWorkbenchPart().getAdapter(ExtensionService.class), "Unable to retrieve extension service");
+		final NamingService namingService = Objects.requireNonNull(getWorkbenchPart().getAdapter(NamingService.class),
+				"Unable to retrieve naming service");
+
+		// This is safe because the constructor requires the part to be an AgeDiagramEditor.
+		final AgeDiagramEditor editor = (AgeDiagramEditor) getWorkbenchPart();
+
+		// Perform modification
+		final DiagramElement dstDiagramElement = getDestinationDiagramElement();
+		final EObject dstBo = getDestinationBusinessObject(dstDiagramElement);
+		final AgeDiagram diagram = DiagamElementUtil.getDiagram(dstDiagramElement);
+		final List<DiagramElement> newDiagramElements = new ArrayList<>();
+		diagram.modify("Paste", m -> {
+			// The modifier will do the actual copying to the diagram elements. It will also copy the business objects
+			// if the copied element includes the business object.
+			final SimpleModifier<EObject> modifier = dstBoToModify -> {
+				newDiagramElements
+				.addAll(copyClipboardContents(dstBoToModify, dstDiagramElement, m, refBuilder, projectProvider, extService,
+						namingService));
+			};
+
+			// If business objects have been copied, then modify the AADL model. Otherwise, just modify the diagram.
+			final boolean anyHaveBo = getCopiedDiagramElements().stream()
+					.anyMatch(de -> de.getCopiedBusinessObject() != null);
+			if (anyHaveBo) {
+				aadlModificationService.modify(dstBo, modifier);
+			} else {
+				modifier.modify(null);
+			}
+
+			// Update the diagram. This will set business objects and update the diagram to be consistent.
+			final AgeFeatureProvider featureProvider = (AgeFeatureProvider) editor.getDiagramTypeProvider()
+					.getFeatureProvider();
+			final DiagramUpdater diagramUpdater = featureProvider.getDiagramUpdater();
+			diagramUpdater.updateDiagram(diagram);
+		});
+
+		// Update selection to match created diagram elements
+		editor.selectDiagramElements(newDiagramElements.toArray(new DiagramElement[newDiagramElements.size()]));
+	}
+
+	/**
+	 * Copies the clipboard contents to the destination business object and diagram element.
+	 * @return the diagram elements which were created. Does not include children of created diagram elements.
+	 */
+	private Collection<DiagramElement> copyClipboardContents(final EObject dstBoToModify,
+			final DiagramElement dstDiagramElement,
+			final DiagramModification m,
+			final ReferenceBuilderService refBuilder, final ProjectProvider projectProvider,
+			final ExtensionService extService, final NamingService namingService) {
+		// Determine the minimum coordinates from the elements whose positions will be copied
+		final Point minCoordinates = getCopiedDiagramElements().stream()
```

Comment:
Should minCoordinates be allowed to be null?

Reply:
Added comment to code describing why it is okay for minCoordinates to be null.